File: standalone/dzen2-hlwm/lua/output.lua
local gmc = require('.gmc')
local common = require('.common')
local helper = require('.helper')

local _M = {}

-- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- -----
-- initialize

-- custom tag names
_M.tag_shows = {'一 ichi', '二 ni', '三 san', '四 shi', 
  '五 go', '六 roku', '七 shichi', '八 hachi', '九 kyū', '十 jū'}

-- initialize variable segment
_M.segment_windowtitle = ''
_M.tags_status = {}

-- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- -----
-- decoration

_M.separator = '^bg()^fg(' .. gmc.color['black'] .. ')|^bg()^fg()'

-- http://fontawesome.io/
_M.font_awesome = '^fn(FontAwesome-9)'

-- Powerline Symbol
_M.right_hard_arrow = '^fn(powerlinesymbols-14)^fn()'
_M.right_soft_arrow = '^fn(powerlinesymbols-14)^fn()'
_M.left_hard_arrow  = '^fn(powerlinesymbols-14)^fn()'
_M.left_soft_arrow  = '^fn(powerlinesymbols-14)^fn()'

-- theme
_M.pre_icon    = '^fg(' .. gmc.color['yellow500'] .. ')' 
              .. _M.font_awesome
_M.post_icon   = '^fn()^fg()'

-- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- -----
-- main

function _M.get_statusbar_text(monitor)
    local text = ''
    
    -- draw tags, zero based
    for index = 0, #(_M.tags_status)-1 do
        text = text .. _M.output_by_tag(monitor, _M.tags_status[index])
    end
    
    -- draw window title    
    text = text .. _M.output_leftside_top()
  
    return text
end

-- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- -----
-- each segments

function _M.output_by_tag(monitor, tag_status)
    local tag_index  = string.sub(tag_status, 2, 2)
    local tag_mark   = string.sub(tag_status, 1, 1)
    local index      = tonumber(tag_index)-- not a zero based array
    local tag_name   = _M.tag_shows[index]

    -- ----- pre tag

    local text_pre = ''
    if tag_mark == '#' then
        text_pre = '^bg(' .. gmc.color['blue500'] .. ')'
                .. '^fg(' .. gmc.color['black'] .. ')'
                .. _M.right_hard_arrow
                .. '^bg(' .. gmc.color['blue500'] .. ')'
                .. '^fg(' .. gmc.color['white'] .. ')'
    elseif tag_mark == '+' then
        text_pre = '^bg(' .. gmc.color['yellow500'] .. ')'
                .. '^fg(' .. gmc.color['grey400'] .. ')'
    elseif tag_mark == ':' then
        text_pre = '^bg()'
                 .. '^fg(' .. gmc.color['white'] .. ')'
    elseif tag_mark == '!' then
        text_pre = '^bg(' .. gmc.color['red500'] .. ')'
                .. '^fg(' .. gmc.color['white'] .. ')'
    else
        text_pre = '^bg()'
                .. '^fg(' .. gmc.color['grey600'] .. ')'
    end

    -- ----- tag by number
    
    -- assuming using dzen2_svn
    -- clickable tags if using SVN dzen
    local text_name = '^ca(1,herbstclient focus_monitor '
                   .. '"' .. monitor .. '" && '
                   .. 'herbstclient use "' .. tag_index .. '")'
                   .. ' ' .. tag_name ..' ^ca()'

    -- ----- post tag

    local text_post = ""
    if (tag_mark == '#') then
        text_post = '^bg(' .. gmc.color['black'] .. ')'
                       .. '^fg(' .. gmc.color['blue500'] .. ')'
                       .. _M.right_hard_arrow
    end

     
    return text_pre .. text_name .. text_post
end

function _M.output_leftside_top()
    local text = ' ^r(5x0) ' .. _M.separator .. ' ^r(5x0) '
               .. _M.segment_windowtitle

    return text
end

-- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- -----
-- setting variables, response to event handler

function _M.set_tag_value(monitor)
    local command = 'herbstclient tag_status ' .. monitor
    local handle = io.popen(command)
    local result = handle:read("*a")
    handle:close() 
        
    local raw = common.trim1(result)  
    _M.tags_status = common.split(raw, "\t")
end

function _M.set_windowtitle(windowtitle)
    local icon = _M.pre_icon .. '' .. _M.post_icon

    if (windowtitle == nil) then windowtitle = '' end
      
    _M.segment_windowtitle = ' ' .. icon ..
        ' ^bg()^fg(' .. gmc.color['grey700'] .. ') ' .. windowtitle
end


-- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- -----
-- return

return _M

File: standalone/dzen2-hlwm/php/output.php
<?php # using PHP7
require_once(__DIR__.'/gmc.php');

# ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----
# initialize

// assuming $ herbstclient tag_status
// 	#1	:2	:3	:4	.5	.6	.7	.8	.9

// custom tag names
const TAG_SHOWS = ['一 ichi', '二 ni', '三 san', '四 shi', 
  '五 go', '六 roku', '七 shichi', '八 hachi', '九 kyū', '十 jū'];

// initialize variable segment
$segment_windowtitle = ''; # empty string
$tags_status         = []; # empty array
$segment_datetime    = ''; # empty string

# ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----
# decoration

const SEPARATOR = "^bg()^fg(".COLOR['black'].")|^bg()^fg()";

// http://fontawesome.io/
const FONT_AWESOME = '^fn(FontAwesome-9)';

// Powerline Symbol
const RIGHT_HARD_ARROW = '^fn(powerlinesymbols-14)^fn()';
const RIGHT_SOFT_ARROW = '^fn(powerlinesymbols-14)^fn()';
const LEFT_HARD_ARROW  = '^fn(powerlinesymbols-14)^fn()';
const LEFT_SOFT_ARROW  = '^fn(powerlinesymbols-14)^fn()';

// theme
const PRE_ICON    = "^fg(".COLOR['yellow500'].")".FONT_AWESOME;
const POST_ICON   = "^fn()^fg()";

# ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----
# main

function get_statusbar_text($monitor)
{
    global $tags_status;
    $text = '';
    
    // draw tags
    foreach ($tags_status as $tag_status) {
        $text .= output_by_tag($monitor, $tag_status);
     }
    
    //# draw date and time
    $text .= output_by_datetime();
    
    // draw window title    
    $text .= output_by_title();
  
    return $text;
}

# ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----
# each segments

function output_by_tag($monitor, $tag_status)
{
    $tag_index  = substr($tag_status, 1, 1);
    $tag_mark   = substr($tag_status, 0, 1);
    $tag_name   = TAG_SHOWS[(int)$tag_index - 1]; # zero based

    # ----- pre tag

    switch ($tag_mark) {
    case "#":
        $text_pre = "^bg(".COLOR['blue500'].")^fg(".COLOR['black'].")"
                  . RIGHT_HARD_ARROW
                  . "^bg(".COLOR['blue500'].")^fg(".COLOR['white'].")";
        break;
    case "+":
        $text_pre = "^bg(".COLOR['yellow500'].")"
                  . "^fg(".COLOR['grey400'].")";       
        break;
    case ":":
        $text_pre = "^bg()^fg(".COLOR['white'].")";
        break;
    case "!":
        $text_pre = "^bg(".COLOR['red500'].")"
                  . "^fg(".COLOR['white'].")";
        break;
    default:
        $text_pre = "^bg()^fg(".COLOR['grey600'].")";
    }

    # ----- tag by number
    
    // assuming using dzen2_svn
    // clickable tags if using SVN dzen
    $text_name = "^ca(1,herbstclient focus_monitor \"${monitor}\" && " 
               . "herbstclient use \"${tag_index}\") ${tag_name} ^ca()";

    # ----- post tag

    if ($tag_mark == '#')
        $text_post = "^bg(".COLOR['black'].")^fg(".COLOR['blue500'].")" 
                   . RIGHT_HARD_ARROW;
    else
        $text_post = "";
     
    return $text_pre . $text_name . $text_post;
}

function output_by_title()
{
    global $segment_windowtitle;
    
    $text  = " ^r(5x0) ".SEPARATOR." ^r(5x0) ";
    $text .= $segment_windowtitle;
    
    return $text;
}

function output_by_datetime()
{
    global $segment_datetime; 
    
    $text  = " ^r(5x0) ".SEPARATOR." ^r(5x0) ";
    $text .= $segment_datetime;
    
    return $text;
}

# ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----
# setting variables, response to event handler

function set_tag_value($monitor)
{
    global $tags_status;

    $raw = shell_exec("herbstclient tag_status $monitor");
    $tags_status = explode("\t", trim($raw));
}

function set_windowtitle($windowtitle)
{
    global $segment_windowtitle;
    $icon = PRE_ICON."".POST_ICON;
      
    $segment_windowtitle = " ${icon} ^bg()"
        . "^fg(".COLOR['grey700'].") $windowtitle";
}

function set_datetime() {
    global $segment_datetime;

    $date_icon = PRE_ICON."".POST_ICON;
    $date_format = '%a %b %d';
    $date_str  = strftime($date_format);
    $date_text = "$date_icon ^bg()^fg(".COLOR['grey700'].") $date_str";

    $time_icon = PRE_ICON."".POST_ICON;
    $time_format = '%H:%M:%S';
    $time_str = strftime($time_format);
    $time_text = "$time_icon ^bg()^fg(".COLOR['blue500'].") $time_str";

    $segment_datetime = "$date_text  $time_text";
}
